codegenerator/cli/src/config_parsing/system_config.rs

```diff
@@ -45,10 +47,327 @@ pub enum Ecosystem {
+                match result {
+                    Ok(val) => Some(val),
+                    Err(_) => None,
+                }
+            }
+        }
+    }
+}
+
+mod interpolation {
+    use anyhow::{anyhow, Result};
+    use regex::{Captures, Regex};
+
+    #[derive(PartialEq)]
+    enum InterpolationResult {
+        DirectSubstitution,
+        InvalidName,
+        DefaultForMissing(String),
+        DefaultForMissingAndEmpty(String),
+    }
+
+    fn parse_capture(inner: &str) -> (String, InterpolationResult) {
+        let (name, result) = match (inner.find(":-"), inner.find('-')) {
+            (Some(pos1), Some(pos2)) if pos1 < pos2 => {
+                let name = &inner[..pos1];
+                let default_value = inner[pos1 + 2..].to_string();
+                (
+                    name,
+                    InterpolationResult::DefaultForMissingAndEmpty(default_value),
+                )
+            }
+            (_, Some(pos)) => {
+                let name = &inner[..pos];
+                let default_value = inner[pos + 1..].to_string();
+                (name, InterpolationResult::DefaultForMissing(default_value))
+            }
+            (Some(pos), _) => {
+                let name = &inner[..pos];
+                let default_value = inner[pos + 2..].to_string();
+                (
+                    name,
+                    InterpolationResult::DefaultForMissingAndEmpty(default_value),
+                )
+            }
+            (None, None) => (inner, InterpolationResult::DirectSubstitution),
+        };
+
+        if name.is_empty()
+            || name.chars().next().map_or(false, |c| c.is_ascii_digit())
+            || !name.chars().all(|c| match c {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '_' => true,
+                _ => false,
+            })
+        {
+            return (name.to_string(), InterpolationResult::InvalidName);
+        }
+
+        (name.to_string(), result)
+    }
+
+    pub fn interpolate_config_variables(
+        config_string: String,
+        mut get_env: impl FnMut(&str) -> Option<String>,
+    ) -> Result<String> {
+        let mut missing_vars = Vec::new();
+        let mut invalid_vars = Vec::new();
+
+        // If we don't have `[^}]` and simpley use `.` in the regex, it will match the last `}` and the rest of the string until the last `}`
+        let re = Regex::new(r"\$\{([^}]*)\}").unwrap();
+        let config_string = re.replace_all(&config_string, |caps: &Captures| {
+            let name = &caps[1];
+            let (name, interpolation_result) = parse_capture(name);
+            if interpolation_result == InterpolationResult::InvalidName {
+                // Wrap invalid vars with quotes to make them more visible in the error message
+                // Don't need to do this for missing ones, because they won't have spaces in the name
+                invalid_vars.push(format!("\"{name}\""));
+                return "".to_string();
+            }
```

Comment:
Should this maybe be put in the match arm?

Also what happens with `DirectSubstitute`? Maybe you could add the last cases for an exhaustiveness enforcement on the match instead of default?

Reply:
I don't want to call `get_env` when the name is invalid.